oak_crypto/src/encryptor.rs

```diff
@@ -230,16 +230,29 @@ pub struct ServerEncryptor {
 }
 
 impl ServerEncryptor {
-    pub fn create(
+    pub fn create<E: EncryptionKeyHandle + ?Sized>(
         serialized_encapsulated_public_key: &[u8],
-        encryption_key_handle: Arc<dyn EncryptionKeyHandle>,
+        encryption_key_handle: &E,
     ) -> anyhow::Result<Self> {
         let recipient_context = encryption_key_handle
             .generate_recipient_context(serialized_encapsulated_public_key)
             .context("couldn't generate recipient crypto context")?;
         Ok(Self::new(recipient_context))
     }
 
+    // TODO(#4311): Merge `create` and `create_async` once there is `async` support in the
+    // Restricted Kernel.
+    pub async fn create_async<E: AsyncEncryptionKeyHandle + ?Sized>(
+        serialized_encapsulated_public_key: &[u8],
```

Comment:
Encapsulated public key shouldn't be a part of the public API, because it will make our clients parse Protobuf encrypted messages themselves.

Getting the public key should be done during `decrypt` function, the same way it was done in the `async` server encryptor

Reply:
(a) This mirrors the behaviour of the sync `create` method. Are you saying the sync `create` should be refactored to mirror the old behaviour of the `create_async`?

(b) I actually disagree about `decrypt` getting the public key and stashing it somewhere. It's weird behaviour; I'd expect decryption to be stateless (as it's the inverse operation of encryption, which _is_ stateless).

We could refactor `create`/`create_async` to take an `EncryptedRequest` as an argument, instead of the raw `serialized_encapsulated_public_key`, but I strongly argue against making decrypt stateful.

Reply:
Yes, the option (a)

I also agree that stateless `decrypt` is more understandable, but because we receive the public key in the client request. So we either need to make our clients parse the encrypted message themselves (which is brittle, because it'll be hard to modify this logic in the future).

Or we have to pass the encrypted message to the `create` function. But in this case it'll need to return decrypted message as well as the encryptor. Or we'll need to again pass the encrypted message to the `decrypt` message afterwards.

So having a stateful `decrypt` function can be a middle-ground for the pure reason of encapsulation.

Reply:
I don't like that middle ground :)

PTAL. I changed `create`-s to take the encrypted message. Yes, you'll have to call `decrypt` on the same message after that, but they are semantically different operations, so I'd argue it captures the intent better this way.

(I guess a real Rusty way would be to create a trait, `IntoEncryptedPublicKey` or something, that both `&[u8; X]`  and `EncryptedRequest` would implement. That way it wouldn't matter whether you have a raw key in the correct format somehow or an encrypted request.)

Reply:
This changes our application SDK that we've already agreed on, so these changes will need to be reflected in the doc and in C++ code as well (to be consistent)
https://github.com/project-oak/oak/blob/d46bcc0d55b6db20b47a76cb84aa37bf9fe73a02/cc/crypto/server_encryptor.h#L45-L46

Also, since both `encrypt` and `decrypt` functions can only be used once (and the plan was to rename this to `ServerSingleUseEncryptor`), the Rusty was would be to have a static `decrypt` function that returns both the decrypted message and the `Encryptor` (which only has a `encrypt` function to encrypt responses)

Reply:
Okay, if you want to enforce that decrypt/encrypt can be used once, then yes, some more refactoring is needed. But in order to get to that kind of API you'd need to get rid of the internal mutable state anyway.

I'll see about refactoring it that way.

Reply:
> Also, since both encrypt and decrypt functions can only be used once (and the plan was to rename this to ServerSingleUseEncryptor), the Rusty way would be to have a static decrypt function that returns both the decrypted message and the Encryptor (which only has a encrypt function to encrypt a single response)

This may even be better than explaining our prod teams that they need to use the encrypted request twice.